src/utils/utils_part.c

```diff
@@ -190,3 +190,79 @@ void ocf_part_set_valid(struct ocf_cache *cache, ocf_part_id_t id,
 		}
 	}
 }
+
+static inline uint32_t ocf_part_evict_size(struct ocf_request *req)
+{
+	uint32_t needed_cache_lines, part_available, cache_lines_to_evict;
+	uint32_t part_occupancy, part_occupancy_debt;
+	uint32_t part_occupancy_limit =
+		ocf_part_get_max_size(req->cache, req->part_id);
+
+	needed_cache_lines = ocf_engine_repart_count(req) +
+		ocf_engine_unmapped_count(req);
+
+	part_occupancy = ocf_part_get_occupancy(req->cache, req->part_id);
+
+	if (part_occupancy_limit >= part_occupancy) {
+		part_available = part_occupancy_limit - part_occupancy;
+		part_occupancy_debt = 0;
+	} else {
+		/* Occupancy is greater than occupancy limit. Evict missing number of
+		 * cachelines */
+		part_occupancy_debt = part_occupancy - part_occupancy_limit;
+		part_available = 0;
+	}
+
+	if (ocf_freelist_num_free(req->cache->freelist) <
+			ocf_engine_unmapped_count(req)) {
+		/* Number of cachelines to insert greater than number of free
+		 * cachelines */
+		if (part_available >= needed_cache_lines) {
+			/* Cache is full, but target's part occupancy limit is not reached
+			   */
+			ocf_req_reset_part_evict(req);
+			cache_lines_to_evict = needed_cache_lines;
+		} else {
+			/* Cache is full and target part reached it's occupancy limit */
+			ocf_req_set_part_evict(req);
+			cache_lines_to_evict = needed_cache_lines - part_available;
+		}
+
+	} else if (part_available < needed_cache_lines) {
+		/* Enough of free cache lines, but partition reached it's occupancy
+		 * limit */
+		cache_lines_to_evict = needed_cache_lines - part_available;
+		ocf_req_set_part_evict(req);
+
+	} else if (part_available >= needed_cache_lines) {
+		/* Enough free cachelines available and they can be assigned to target
+		 * partition */
+		cache_lines_to_evict = 0;
+
+	}
+
+	return cache_lines_to_evict + part_occupancy_debt;
```

Comment:
Maybe we should consider limiting size of eviction induced by occupancy debt to some fixed constant? This would limit the overhead of occupancy debt on a single I/O